internal/action/render.go

```diff
@@ -0,0 +1,235 @@
+package action
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/operator-framework/operator-registry/internal/declcfg"
+	"github.com/operator-framework/operator-registry/internal/property"
+	"github.com/operator-framework/operator-registry/pkg/containertools"
+	"github.com/operator-framework/operator-registry/pkg/image"
+	"github.com/operator-framework/operator-registry/pkg/image/containerdregistry"
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	"github.com/operator-framework/operator-registry/pkg/sqlite"
+)
+
+type Render struct {
+	Refs     []string
+	Registry image.Registry
+}
+
+func nullLogger() *logrus.Entry {
+	logger := logrus.New()
+	logger.SetOutput(ioutil.Discard)
+	return logrus.NewEntry(logger)
+}
+
+func (r Render) Run(ctx context.Context) (*declcfg.DeclarativeConfig, error) {
+	if r.Registry == nil {
+		reg, err := r.createRegistry()
+		if err != nil {
+			return nil, fmt.Errorf("create registry: %v", err)
+		}
+		defer reg.Destroy()
+		r.Registry = reg
+	}
+
+	var cfgs []declcfg.DeclarativeConfig
+	for _, ref := range r.Refs {
+		var (
+			cfg *declcfg.DeclarativeConfig
+			err error
+		)
+		if stat, serr := os.Stat(ref); serr == nil && stat.IsDir() {
+			cfg, err = declcfg.LoadDir(ref)
+		} else {
+			cfg, err = r.imageToDeclcfg(ctx, ref)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("render reference %q: %v", ref, err)
+		}
+		renderBundleObjects(cfg)
+		cfgs = append(cfgs, *cfg)
+	}
+
+	return combineConfigs(cfgs), nil
+}
+
+func (r Render) createRegistry() (*containerdregistry.Registry, error) {
+	cacheDir, err := os.MkdirTemp("", "render-registry-")
+	if err != nil {
+		return nil, fmt.Errorf("create tempdir: %v", err)
+	}
+
+	reg, err := containerdregistry.NewRegistry(
+		containerdregistry.WithCacheDir(cacheDir),
+		containerdregistry.WithLog(nullLogger()),
```

Comment:
These logs are pretty verbose even in the "happy" case. For example, https://github.com/operator-framework/operator-registry/blame/c95be3c757ffedf26e8ceb89b6d9462ced1604c6/pkg/image/containerdregistry/registry.go#L167 produces a pretty long JSON log string that looks similar to the `docker inspect` output.

Is it reasonable to expect that any messages that need to be surfaced to the user are returned as `error`s rather than logged?